### PR TITLE
Add registration/authentication extensions for cloud-assisted BLE

### PR DESCRIPTION
**Pull request by kpaulh**
This is a counterpart to https://github.com/fido-alliance/fido-2-specs/pull/529/, the FIDO2 cloud-assisted BLE PR. We're not quite sure where the extension should live - FIDO2 or WebAuthN. We're leaning towards WebAuthN, since it is relevant to RPs, but hoping others have some thoughts.

Also, this PR currently has references to sections in the FIDO2 spec that I can either replace with complete explanations or update links to once they're available in the FIDO spec. We figured we would get eyes on it now and clean up along the way.

Note that since this is the first registration extension, this PR also adds RegistrationExtensionsClientInputs/Outputs.

***
<a href="https://pr-preview.s3.amazonaws.com/kpaulh/webauthn/pull/909.html" title="Last updated on May 1, 2019, 11:44 PM UTC (63be77d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/909/5d11e62...kpaulh:63be77d.html" title="Last updated on May 1, 2019, 11:44 PM UTC (63be77d)">Diff</a>

### CODE REVIEW

**Comment by nadalin** [comment]
@kpaulh Will review @ the FIDO Plenary

File: index.bs [hunk 1]
```diff
@@ -1918,6 +1918,33 @@ SHOULD be aborted.
...
+
+
+## Registration Extensions Client Outputs (typedef <dfn>RegistrationExtensionsClientOutputs</dfn>) ## {#iface-registration-extensions-client-outputs}
+
+<xmp class="idl">
+    dictionary RegistrationExtensionsClientOutputs {
+    };
+</xmp>
+
+This is a dictionary containing the [=client extension output=] values for zero or more WebAuthn extensions, as defined in [[#extensions]].
```
Comment (selfissued) [inline index.bs]: Delete RegistrationExtensionsClientOutputs.  Use AuthenticationExtensionsClientOutputs instead.

File: index.bs [hunk 2]
```diff
@@ -1918,6 +1918,33 @@ SHOULD be aborted.
...
     See [WHATWG HTML WG Issue #2711](https://github.com/whatwg/html/issues/2711) for more details. 
 
+## Registration Extensions Client Inputs (typedef <dfn>RegistrationExtensionsClientInputs</dfn>) ## {#iface-registration-extensions-client-inputs}
+
+<xmp class="idl">
+    dictionary RegistrationExtensionsClientInputs {
+    };
+</xmp>
+
+This is a dictionary containing the [=client extension input=] values for zero or more WebAuthn extensions, as defined in [[#extensions]].
```
Comment (selfissued) [inline index.bs]: Delete RegistrationExtensionsClientInputs.  Use AuthenticationExtensionsClientInputs instead.

File: index.bs [hunk 3]
```diff
@@ -1918,6 +1918,33 @@ SHOULD be aborted.
...
+This is a dictionary containing the [=client extension output=] values for zero or more WebAuthn extensions, as defined in [[#extensions]].
+
+
+## Registration Extensions Authenticator Inputs (typedef <dfn>RegistrationExtensionsAuthenticatorInputs</dfn>) ## {#iface-registration-extensions-authenticator-inputs}
+
+<xmp class="idl">
+    typedef record<DOMString, DOMString> RegistrationExtensionsAuthenticatorInputs;
+</xmp>
+
+This is a dictionary containing the [=authenticator extension input=] values for zero or more WebAuthn extensions, as defined in [[#extensions]].
```
Comment (selfissued) [inline index.bs]: Delete RegistrationExtensionsAuthenticatorOutputs.  Use AuthenticationExtensionsAuthenticatorOutputs instead.

File: index.bs [hunk 4]
```diff
@@ -4303,6 +4330,112 @@ as candidates to be employed in a [=registration=] ceremony.
...
+
+: Client extension input
+:: A Javascript object defined as follows:
+    <xmp class="idl">
+    dictionary CableRegistrationData {
+      required sequence<long> versions;
+      required BufferSource rpPublicKey;
+    };
+
+    partial dictionary RegistrationExtensionsClientInputs {
```
Comment (selfissued) [inline index.bs]: Registration -> Authentication

File: index.bs (hunk 1 above)
Comment (kpaulh) [reply index.bs]: Done! sorry for the confusion.

File: index.bs (hunk 2 above)
Comment (kpaulh) [reply index.bs]: Done

File: index.bs (hunk 3 above)
Comment (kpaulh) [reply index.bs]: Done

File: index.bs (hunk 4 above)
Comment (kpaulh) [reply index.bs]: and Done

**Comment by nadalin** [comment]
@jcjones @akshayku @equalsJeffH  Please review

**Comment by equalsJeffH** [comment]
Just to note wrt
> We're not quite sure where the[se] extension[s] should live - FIDO2 or WebAuthN

Just to note, presently these extension definitions are also contained in https://github.com/fido-alliance/fido-2-specs/pull/529/ [ which is for the CTAP spec ]. If we decide to land them here in the WebAuthn spec, we will need to update https://github.com/fido-alliance/fido-2-specs/pull/529/ such that they (specifically only the extension definitions) are elided (from the latter CTAP-specific PR).

File: index.bs [hunk 5]
```diff
@@ -4892,6 +4892,112 @@ as candidates to be employed in a [=registration ceremony=].
...
+    CableRegistration = {
+                           version: int,
+                           maxVersion: int,
+                           authenticatorPublicKey: bytes,
+                       }
+    ```
+
+During <code>authenticatorGetAssertion()</code>, the RP provides the
+client with the necessary session data via the following extension.
+For details refer to Sections [#cable-eid] and [#cable-encryption].
```
Comment (jcjones) [inline index.bs]: I don't think these anchors exist?

**Comment by nadalin** [comment]
We will leave open until the CTAP work is implemented and make sure this works with passwordless world

File: index.bs [hunk 6]
```diff
@@ -4892,6 +4892,112 @@ as candidates to be employed in a [=registration ceremony=].
...
+      required sequence<long> versions;
+      required BufferSource rpPublicKey;
+    };
+
+    partial dictionary AuthenticationExtensionsClientInputs {
+      CableRegistrationData cableRegistration;
+    };
+    </xmp>
+
+    Versions is a list of protocol versions supported by the relying party.
```
Comment (ynojima) [inline index.bs]: Could you elaborate the `version` ? What value is supported here?

File: index.bs [hunk 7]
```diff
@@ -4892,6 +4892,112 @@ as candidates to be employed in a [=registration ceremony=].
...
+
+    The rpPublicKey is an ECDH P-256 public key.
```
Comment (ynojima) [inline index.bs]: What format is used for encoding `rpPublicKey`?

File: index.bs [hunk 8]
```diff
@@ -4892,6 +4892,112 @@ as candidates to be employed in a [=registration ceremony=].
...
+    partial dictionary AuthenticationExtensionsClientInputs {
+      sequence<CableAuthenticationData> cableAuthentication;
+    };
+    </xmp>
+
+    The version is the protocol version used during registration.
+
+    The clientEid is a 16-byte EID advertised by the Client.
+
+    The authenticatorEid is a 16-byte EID advertised by the Authenticator.
```
Comment (ynojima) [inline index.bs]: How does the RP retrieve the EID advertised by the Authenticator?

**Comment by ynojima** [comment]
It seems `AuthenticatorTransport` enum should have `cable` value.

https://github.com/chromium/chromium/blob/e4aee4ff3c5981448f8b9cbc9bc948f005d86392/third_party/blink/renderer/modules/credentialmanager/public_key_credential_descriptor.idl#L11

File: index.bs (hunk 6 above)
Comment (arnar) [reply index.bs]: For now only "1", but the field exists so that we can make changes/fixes to the protocol if needed.

**Comment by equalsJeffH** [comment]
we are closing this because we are hoping it will not be needed.

**Comment by equalsJeffH** [comment]
at 2020-02-26 meeting, we decided to close this because we are sure it wont be needed.

**Comment by zbot473** [comment]
@equalsJeffH What is the standard for this then?
How will smartphone auth be done, or is that not the spec's job?

**Pull request closed without merging**